src/metrics.rs

```diff
@@ -0,0 +1,308 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+use histogram::Histogram;
+
+use tokio::sync::{mpsc, oneshot};
+use tokio::time;
+use tokio_util::sync::CancellationToken;
+
+use ringbuffer::{AllocRingBuffer, RingBuffer};
+
+use std::time::Duration;
+
+use log::*;
+
+#[derive(Clone, Debug)]
+pub enum AggregatedMetricsMessages {
+    AggWindow {
+        messages_in_window: u64,
+        window_epoch: u32,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub enum ClientMessages {
+    MessageProcessed { client_id: usize },
+    ClientFailed { client_id: usize },
+}
+
+#[derive(Clone, Debug)]
+pub struct WindowConfig {
+    pub window_start: time::Instant,
+    pub window_duration: Duration,
+}
+
+pub struct MetricsAggregator {
```

Comment:
So there is one of these per producer now? I don't understand why we need a channel then to receive messages? Are you planning on sharing them across multiple instances?

Reply:
Yeah, there is one per producer/consumer now. I would've use a single-producer-single-consumer channel, but Rust/Tokio doesn't seem to have one included in their standard libs.

You can sort of imagine `MetricsAggregator` to be similar to a housekeeping task we have in RP. In RP we schedule a task to run at set intervals. This task has references to the class it's running for and we protect those references using a gate or similar.  

However, in Rust due how lifetimes work and that tasks aren't guaranteed to run on the same thread we need to either use an atomic referenced counted pointer with a mutex or a channel to share data. Hence the channel.

Reply:
but if it's a single one per producer/consumer why can we not just make it a local/member?

Reply:
Then we would need logic in the producers/consumers to ensure that the collected metrics are sent every `window_duration` even if they are empty/zero. It was a lot simpler to have a task that loops on a future created by `time::Interval` to do this.

Reply:
Indeed that could be another way to go about it. And you would only need one `time::Interval` over all producers/consumers unlike the current approach of needing one per producer/consumer. The only issue I suppose would be that you would still need an `Arc<Mutex<...>>` around each set of metrics being collected per producer/consumer as one or more threads could be accessing said metrics.

Reply:
I did actually mean doing it on a per producer/consumer basis. But yeah maybe I am overthinking stuff.

Reply:
It'd definitely be possible and beyond having to partially re-write the producer/consumer it may be the better approach. I can switch to it if you want. I don't see the need to though.